File: src/command/cpp/command.cc
#include <iostream>
#include "command.h"

using namespace std;

void Reciever::Action()
{
	cout << "Do action" << endl;
}

Read_Command::Read_Command(Reciever* rev)
{
	this->m_pRev = rev;
}

void Read_Command::Execute()
{
	cout << "Read command..." << endl;
	m_pRev->Action();
}

Read_Command::~Read_Command()
{
	delete m_pRev;
}

Write_Command::Write_Command(Reciever* rev)
{
	this->m_pRev = rev;
}

void Write_Command::Execute()
{
	cout << "Write command..." << endl;
	m_pRev->Action();
}

Write_Command::~Write_Command()
{
	delete m_pRev;
}

Invoker::Invoker(Command* cmd)
{
	m_pCmd = cmd;
}

void Invoker::Notify()
{
	list<Command*>::iterator it = m_lCmdList.begin();
	for(; it != m_lCmdList.end(); ++it) {
		m_pCmd = *it;
		m_pCmd->Execute();
	}
}

void Invoker::AddCmd(Command* cmd)
{
	m_lCmdList.push_back(cmd);
}

void Invoker::DelCmd(Command* cmd)
{
	m_lCmdList.remove(cmd);
}

Invoker::~Invoker()
{
	delete m_pCmd;
}

File: src/command/php/command.php
<?php
/**
 * command pattern
 */

interface Command {
	public function execute();
}

class ConcreteCommand implements Command {
	private $_receiver;

	public function __construct(Receiver $receiver) {
		$this->_receiver = $receiver;
	}

	public function execute() {
		$this->_receiver->action();
	}
}

class Receiver {
	/* 接收者名称 */
	private $_name;

	public function __construct($name) {
		$this->_name = $name;
	}

	public function action() {
		printf("%s -> %s\n", $this->_name, __FUNCTION__);
	}
}

class Invoker {
	private $_command;

	public function __construct(Command $command) {
		$this->_command = $command;
	}

	public function action() {
		$this->_command->execute();
	}
}

class Client {
	public static function main() {
		$receiver = new Receiver("John");
		$command = new ConcreteCommand($receiver);
		$invoker = new Invoker($command);
		$invoker->action();
		$receiver = null;
		$command = null;
		$invoker = null;
	}
}

Client::main();

?>
